File: Problems/LifeTree.java
import java.util.*;


public class HelloWorld {
    public static void main(String[] args) {
      List<Relation> input = new ArrayList<>();
      input.add(new Relation("animal", "mammal"));
      input.add(new Relation("animal", "bird"));
      input.add(new Relation("lifeform", "animal"));
      input.add(new Relation("cat", "lion"));
      input.add(new Relation("mammal", "cat"));
      input.add(new Relation("animal", "fish"));
      
      printTree(input);
    }
    
    public static void printTree(List<Relation> rs) {
      List<String> newList;
      Map<String, List<String>> map = new HashMap();
      Set<String> children = new HashSet();
      
      for(Relation relation: rs) {
        children.add(relation.child);
        if(map.containsKey(relation.parent)) {
          map.get(relation.parent).add(relation.child);
        } else {
          newList = new ArrayList<>();
          newList.add(relation.child);
          map.put(relation.parent, newList);
        }
      }
      String root = "";
      for(Map.Entry<String, List<String>> entry: map.entrySet()) {
        if(!children.contains(entry.getKey())) {
          root = entry.getKey();
          break;
        }
      }
      dfs(root, 0, map);
    }
    
    public static void dfs(String root, int level, Map<String, List<String>> map) {
      for(var i=0; i<level; i++) {
        System.out.print("\t");
      }
      System.out.println(root);
      List<String> children;
      if(map.containsKey(root)) {
        children = map.get(root);
      } else {
        return;
      }
      for(String child: children) {
        dfs(child, level+1, map);
      }
    }
}


class Relation {
  String parent;
  String child;
  public Relation(String parent, String child) {
    this.parent = parent;
    this.child = child;
  }
}

File: Problems/LifeTree.js
/*
    (GPI)
    Write a function displayTree(), which prints a given tree to stdout.

    - Parameter of the function is a set of relations (parent -> child).
    - Each string found anywhere in the input represents a unique node.
    - Each parent can have many children.
    - Input list may contain relations in any order.
    - Order in which pairs appear in the input list determines the node order to its sibilings.

    Input:
    ------
    
    class Relation {
        constructor(parent, child) {
            this.parent = parent;
            this.child = child;
        }
    }

    input.push(new Relation("animal", "mammal"));
    input.push(new Relation("animal", "bird"));
    input.push(new Relation("lifeform", "animal"));
    input.push(new Relation("cat", "lion"));
    input.push(new Relation("mammal", "cat"));
    input.push(new Relation("animal", "fish"));

    printTree(input);

    Output:
    -------

    lifeform
        animal
            mammal
                cat
                    lion
            bird
            fish

*/

class Relation {
    constructor(parent, child) {
        this.parent = parent;
        this.child = child;
    }
}

var input = [];

input.push(new Relation("animal", "mammal"));
input.push(new Relation("animal", "bird"));
input.push(new Relation("lifeform", "animal"));
input.push(new Relation("cat", "lion"));
input.push(new Relation("mammal", "cat"));
input.push(new Relation("animal", "fish"));

function printTree(list) {
    var hashMap = {}; // hashset to adjacenyList with parent and childs
    var result; // temp array to push childs to its respective parent
    var children = new Set(); // the adjacenyList which is answer to print

    var tree = [];

    for(var val of list) {
        children.add(val.child);
        if(hashMap[val.parent]) {
            hashMap[val.parent].push(val.child);
        } else {
            result = [];
            result.push(val.child);
            hashMap[val.parent] = result; 
        }
    }
          
    function dfs(root, level, map) {        

        tree.push(root);

        var childs = [];
        if(map[root]) {
            childs = map[root];
        } else return;

        for(var val of childs) {
            dfs(val, level+1, map);
        }
    }

    // finding the root
    let root = "";
    for(var key in hashMap) {
        if(!children.has(key)) {
            root = key;
            break;
        }
    }
    
    dfs(root, 0, hashMap);

    return tree;
}

console.log(printTree(input));

// REPRESENTAION OF THE TREE
/*
            lifeform
               |
             animal
            /  \  \
        mammal bird fish
          |
         cat
          |
         lion
*/